### ISSUE TITLE
@okta/okta-react - <SecureRoute/>  - fails to refresh tokens - workaround needed

### ISSUE THREAD

**Issue by mrt123** [okta/okta-oidc-js#254]
My route is secured with SecureRoute component as below:
```
<Security issuer={myIssuer} client_id={myClientId} redirect_uri={myRedirectUri}>
    <SecureRoute path="/protected" component={MyComponent}/>
<\Security>
```
Without any modification to pure MyComponent, when accessing /protected I get taken through OKTA login cycle as per my config and everything is fine. However when I access the app the next day, /protected gets accessed without a problem but my REST API calls return 401. Upon page refresh API calls return 200, while no redirect to OKTA was made in the process.

This leads me to believe access token was expired. I know you guys have issues with tokens not being refreshed on time. 

**I'd like to know if there's any elegant workaround for this issue until this is solved? I can live with a refresh of the page if token is in fact expired.**

**Comment by hotgazpacho** [comment]
What it looks like is happening is that when the `SecureRoute` calls `auth.isAuthenticated()` here: https://github.com/okta/okta-oidc-js/blob/master/packages/okta-react/src/Auth.js#L53-L57. That looks like it is merely checking for the presence of an access token or a refresh token, but not whether they have expired or not. Digging into the `tokenManager.get()` method here: https://github.com/okta/okta-auth-js/blob/master/lib/TokenManager.js#L97-L100 seems to confirm that. Now, the `tokenManager` is supposed to be handling expiring of tokens automatically, but I suspect that there's a timing issue between when the `SecureRoute` component asks for the tokens, and when the `tokenManager` is able to enter it's event loop to check for expired tokens.

It seems this is a symptom of an issue the team is tracking in okta/okta-auth-js#125

**Comment by mrt123** [comment]
@hotgazpacho do we have any elegant workaround that I could use until this gets fixed? I tried immediate page reload if my API returns 401, but that doesn't seem to work.

**Comment by hotgazpacho** [comment]
Just spit-balling here, but if you can get a hold of the `Auth` instance at app initialization time, you could replace [the `isAuthenticated()` method](https://github.com/okta/okta-oidc-js/blob/master/packages/okta-react/src/Auth.js#L53-L57) with your own implementation that not only asks for the tokens, but then also checks their `exp` claim to see if the are expired.

I'm running into the same issue on one of my own projects, but I haven't attempted a fix yet.

**Comment by hotgazpacho** [comment]
Here's what I ended up doing. I build the `auth` instance myself, and pass it as a prop to the `Security` component.

```js
const createAuth = (history, oktaConfig) => {
  const auth = new Auth({
    history,
    ...oktaConfig
  });

  // Overwrite the isAuthenticated function because of a bug in the okta-oidc-js
  // library (https://github.com/okta/okta-auth-js/issues/125).
  // TL;DR: this method assumes that the tokenManager will only ever have non-expired
  // tokens, but this is demonstrably false if you come back to an app later,
  // after the token as expired, but before tokenManager has had a chance to remove it,
  // meaning it is possible to view components protected by a SecureRoute with an
  // expired token.
  const containsAuthTokens = /id_token|access_token|code/;
  auth.isAuthenticated = async () => {
    // If there could be tokens in the url
    // eslint-disable-next-line no-restricted-globals
    if (location && location.hash && containsAuthTokens.test(location.hash)) return null;
    const accessToken = auth._oktaAuth.tokenManager.get('accessToken');
    return !!(accessToken && accessToken.expiresAt * 1000 > Date.now());
  };

  return auth;
}
```

**Comment by mrt123** [comment]
Nice @hotgazpacho 

In my case since I'm using '@okta/okta-react' to secure routes, and '@okta/okta-auth-js' to secure my API, I've added the following code to provoke login through redirect (hence it obtains new token). I have my auth server set to expire token after 8 hours, and frankly to me getting user re-logged after token expiration is the desired behaviour.


```
	const oktaResponse = oktaAuth.tokenManager.get('accessToken') // gets locally stored token (stored earlier by @okta/okta-react) 
	const tokenExpiresAt = oktaResponse !== undefined ? oktaResponse.expiresAt * 1000 : 0
	const tokenExpired = tokenExpiresAt < new Date().getTime()

	if (tokenExpired) {
		oktaAuth.token.getWithRedirect({responseType: 'token'})
	}
	else {
		const tokenString = oktaResponse === undefined ? '' : oktaResponse.accessToken
		const headers = {
			...extraHeaders,
			Authorization: 'Bearer ' + tokenString
		}

		return fetch(url, {headers}).then(r => r.json());
	}
```

**Comment by robertjd** [comment]
Hi everyone, this issue should now be resolved if you upgrade to `@okta/okta-react@1.0.3`, which now depends on version 2 of AuthJS.  Ensure that you have `@okta/okta-auth-js@2.0.1` in your tree after the upgrade.

I'm going to close this issue, but please re-open if needed.  Thanks everyone for the discussion, and thank you @hotgazpacho for sharing the workaround while we fixed this!

**Comment by rodoabad** [comment]
@robertjd not sure if this has been completely fixed since the latest version `1.1.0` still has this issue.

https://github.com/okta/okta-oidc-js/issues/266
